Auto-offline for `conda_install`

smarie opened the pull request:
Changelog (let me know if I should commit it in the changelog file):

 - Offline mode is now auto-detected by default by `conda_install`. This allows users to continue executing nox sessions on already installed environments. This behaviour can be disabled by setting `auto_offline=False`. Fixes #313

theacodes commented:
@tswast are you willing to take a look at this as our resident conda expert?

theacodes commented on setup.py:
```diff
@@ -54,6 +54,7 @@
         "colorlog>=2.6.1,<5.0.0",
         "py>=1.4.0,<2.0.0",
         "virtualenv>=14.0.0",
+        "requests",
```
I am very against adding new dependencies in nox. Can we think of a way to do this without taking on new dependencies?

smarie replied:
I was afraid of this :)
well, there are two paths here:

 - either we remove `requests` from the dependencies BUT we try to import it when the offline-checking procedure starts, and we do not find it we fallback to the legacy behaviour (no online/offline checks)

 - or we write the equivalent code using stdlib only. The "only" issue with this is that it might be hard to get a complete equivalent feature supporting http proxy settings and other complexities that `requests` handle for us.

Since `requests` is available in most python environments today I would tend to prefer option 1 as the risk of doing the second one wrong is high. What do you think?

smarie replied:
Actually I'll still give it a go as apart from proxies it seems quite ok. Let me try

theacodes commented:
Could we try with dns queries only?

On Fri, Jun 19, 2020, 5:38 AM smarie <notifications@github.com> wrote:

> *@smarie* commented on this pull request.
> ------------------------------
>
> In setup.py
> <https://github.com/theacodes/nox/pull/314#discussion_r442815549>:
>
> > @@ -54,6 +54,7 @@
>          "colorlog>=2.6.1,<5.0.0",
>          "py>=1.4.0,<2.0.0",
>          "virtualenv>=14.0.0",
> +        "requests",
>
> Actually I'll still give it a go as apart from proxies it seems quite ok.
> Let me try
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/theacodes/nox/pull/314#discussion_r442815549>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAB5I437672OSE5WNIKCRH3RXNL3XANCNFSM4MNDQ35A>
> .
>

smarie commented:
No, this is not sufficient: if only using dns check, a user sitting behind a corporate proxy would wrongly get detected as offline I think :(

I modified the PR to get rid of `requests`... only to find out that I still use the lower level `urllib3` which is **still** not in the stdlib...

I'll stop investigating for today - I'll have another look next week.

theacodes commented:
Thanks! & no rush, ofc.

On Fri, Jun 19, 2020 at 9:41 AM smarie <notifications@github.com> wrote:

> No, this is not sufficient: if only using dns check, a user sitting behind
> a corporate proxy would wrongly get detected as offline I think :(
>
> I modified the PR to get rid of requests... only to find out that I still
> use the lower level urllib3 which is *still* not in the stdlib...
>
> I'll stop investigating for today - I'll have another look next week.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/theacodes/nox/pull/314#issuecomment-646738591>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAB5I43IF7XSACROHJKSDFDRXOILFANCNFSM4MNDQ35A>
> .
>

smarie commented:
Hi @theacodes I found a way to do this with stdlib only but... eventually I removed it and reverted to the dns check only :)

Indeed, I found a real-life situation where a failing HTTP query (even with `requests`) is not enough to indicate that the user is offline. Assuming offline when we are not 100% certain is highly undesirable, so I rather remove all this pseudo-intelligence.

The situation is the following: a user sits behind a corporate proxy, and his HTTP proxy is correctly configured for `conda` using the `.condarc` `proxy_servers` section, while not being correctly configured in the `http_proxy` / `https_proxy` / `no_proxy` OS environment variables used by all other tools including python `urllib.requests.getproxies` or `requests`. In this situation, our "pseudo-intelligent" method would be returning a false "offline" diagnostic, which is of course highly undesirable.

I wrote a detailed explanation here in case you're interested: https://stackoverflow.com/a/62486343/7262247

theacodes commented:
Thanks!
